compiler/rustc_middle/src/infer/unify_key.rs

```diff
@@ -43,18 +46,8 @@ impl<'tcx> UnifyKey for RegionVidKey<'tcx> {
 impl<'tcx> UnifyValue for UnifiedRegion<'tcx> {
     type Error = NoError;
 
-    fn unify_values(value1: &Self, value2: &Self) -> Result<Self, NoError> {
-        Ok(match (value1.0, value2.0) {
-            // Here we can just pick one value, because the full constraints graph
-            // will be handled later. Ideally, we might want a `MultipleValues`
-            // variant or something. For now though, this is fine.
-            (Some(_), Some(_)) => *value1,
-
-            (Some(_), _) => *value1,
-            (_, Some(_)) => *value2,
-
-            (None, None) => *value1,
-        })
+    fn unify_values(value1: &Self, value2: &Self) -> Result<Self, Self::Error> {
+        Ok(cmp::min_by_key(*value1, *value2, |val| (val.universe, val.value.is_none())))
```

Comment:
this can pull down the universe of known regions which seems wrong :thinking: I would expect us to have to eagerly error here if we relate with a known region

Reply:
The idea is that unification should always be infallible, the constraint graph will eventually emit a universe error in this case.

The only invariant we should maintain in the unification table is that an existential cannot resolve to a region in a higher universe (plus one nice thing to have is to prefer universals over existential within the same universe, but that's not necessary for soundness).